src/Avalonia.FreeDesktop/DBusSystemDialog.cs

```diff
@@ -0,0 +1,120 @@
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Avalonia.Controls;
+using Avalonia.Controls.Platform;
+using Avalonia.Dialogs;
+using Avalonia.Logging;
+using Tmds.DBus;
+
+namespace Avalonia.FreeDesktop
+{
+    internal class DBusSystemDialog : ISystemDialogImpl
+    {
+        private readonly IFileChooser? _fileChooser;
+        private bool _isDbusAvailable;
+
+        internal DBusSystemDialog()
+        {
+            _fileChooser = DBusHelper.Connection?.CreateProxy<IFileChooser>("org.freedesktop.portal.Desktop", "/org/freedesktop/portal/desktop");
+            _isDbusAvailable = _fileChooser is not null;
+        }
+
+        public async Task<string[]?> ShowFileDialogAsync(FileDialog dialog, Window parent)
+        {
+            if (!_isDbusAvailable)
+                return await dialog.ShowManagedAsync(parent);
+            try
+            {
+                return await ShowNativeFileDialogAsync(dialog, parent);
+            }
+            catch (Exception e)
+            {
+                Logger.TryGet(LogEventLevel.Error, LogArea.X11Platform)?.Log(this, e.Message);
+                _isDbusAvailable = false;
+                return await dialog.ShowManagedAsync(parent);
+            }
+        }
+
+        public async Task<string?> ShowFolderDialogAsync(OpenFolderDialog dialog, Window parent)
+        {
+            if (!_isDbusAvailable)
+                return await dialog.ShowManagedAsync(parent);
+            try
+            {
+                return await ShowNativeFolderDialogAsync(dialog, parent);
+            }
+            catch (Exception e)
+            {
+                Logger.TryGet(LogEventLevel.Error, LogArea.X11Platform)?.Log(this, e.Message);
+                _isDbusAvailable = false;
+                return await dialog.ShowManagedAsync(parent);
+            }
+        }
+
+        private async Task<string[]?> ShowNativeFileDialogAsync(FileDialog dialog, Window parent)
+        {
+            var parentWindow = $"x11:{parent.PlatformImpl!.Handle.Handle.ToString("X")}";
+            ObjectPath objectPath;
+            var options = new Dictionary<string, object>();
+            if (dialog.Filters is not null)
+                options.Add("filters", ParseFilters(dialog));
+
+            switch (dialog)
+            {
+                case OpenFileDialog openFileDialog:
+                    options.Add("multiple", openFileDialog.AllowMultiple);
+                    objectPath = await _fileChooser!.OpenFileAsync(parentWindow, openFileDialog.Title ?? string.Empty, options);
+                    break;
+                case SaveFileDialog saveFileDialog:
+                    if (saveFileDialog.InitialFileName is not null)
+                        options.Add("current_name", saveFileDialog.InitialFileName);
+                    if (saveFileDialog.Directory is not null)
+                        options.Add("current_folder", Encoding.UTF8.GetBytes(saveFileDialog.Directory));
```

Comment:
I haven't tested this PR on linux yet, but is it correct? For "current_name" you pass a string, but for "current_folder" you pass an array of bytes. Or does it work in either way?

Reply:
As the [documentation](https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.FileChooser) states:

> current_name s 
>  Suggested filename.

> current_folder ay
> Suggested folder to save the file in. The byte array is expected to be null-terminated.

I dont't know why it designed like that, but this is what the documentation says.